lib/tpm2_auth_util.c

```diff
@@ -90,6 +91,29 @@ static tool_rc start_hmac_session(ESYS_CONTEXT *ectx, TPM2B_AUTH *auth,
         return tool_rc_general_error;
     }
 
+    if (ectx) {
+        uint32_t i;
+        tool_rc tmp_rc;
+        TPM2_CAP capability = TPM2_CAP_PCRS;
+        UINT32 property = 0;
+        UINT32 count = TPM2_MAX_TPM_PROPERTIES;
+        TPMS_CAPABILITY_DATA *capability_data = NULL;
+        TPMI_ALG_HASH hash = tpm2_session_data_get_authhash(d);
+
+        tmp_rc = tpm2_capability_get(ectx, capability, property, count, &capability_data);
+        if (tmp_rc == tool_rc_success) {
+            for (i = 0; i < capability_data->data.assignedPCR.count; i++) {
+                if(capability_data->data.assignedPCR.pcrSelections[i].hash == hash) {
+                    break;
+                }
+            }
+            if(capability_data->data.assignedPCR.count > 0 && i == capability_data->data.assignedPCR.count) {
+                tpm2_session_set_authhash(d, capability_data->data.assignedPCR.pcrSelections[0].hash);
+            }
+            free(capability_data);
+        }
+    }
+
```

Comment:
@mayuanchenma what is the purpose of setting the authhash in this way?

Reply:
tpm2_session_data_new set SHA256 algorithm in default, if a TPM do not support SHA256, it will cause a TPM error. In the way can check the hash from the TPM dynamically.

Reply:
@mayuanchenma, I see the problem. But I am skeptical of this solution as it may not scale based on how pcr allocation happens. As in, when only SHA1 PCR banks are turned on we might end up with a session using a weaker cipher when SHA256 algorithm was supported.

Nevertheless SHA256 being the only fixed is definitely a problem. However, would it better to get capability to learn the supported algorithms and then choose one with biggest block size if ectx is available and if not default to SHA256?

Reply:
@williamcroberts your thoughts?

Reply:
Why does which PCR banks are enabled affect supported algorithms for TPM2_StartAuthSession? Support for algorithm SHA256 is always enabled in a TPM2.0 device per the PC spec:
  - https://trustedcomputinggroup.org/wp-content/uploads/TCG_PC_Client_Platform_TPM_Profile_PTP_Specification_Family_2.0_Revision_1.3v22.pdf

Reply:
Indeed, the algorithms of the bank is reports, what goes missing in cap after pcrallocation is the pcrs themselves if they have been disabled.
@mayuanchenma can you please dump the output for `tpm2_getcap pcrs` from a TPM that supports SM algs?

Reply:
@idesai  "However, would it better to get capability to learn the supported algorithms and then choose one with biggest block size if ectx is available and if not default to SHA256?", it seems to be a better solution.